### PR TITLE
BUG: Infer compression by default in read_fwf() (#22199)

### LINKED ISSUE #22199: read_fwf() function does not infer compression by default
#### Code Sample, a copy-pastable example if possible
```python
import io
import gzip
import os
import pandas as pd

fwf = b"""a b c\n1 2 3\n4 5 6"""
with gzip.open('file.gz', 'wb') as gz:
    gz.write(fwf)
print(pd.read_fwf('file.gz', compression='infer'))
print(pd.read_fwf('file.gz'))
os.remove('file.gz')
```
#### Problem description
The pandas documentation indicates that read_fwf() (as with read_csv()) infers a file's compression type based on its extension _by default_, but it does not.

#### Expected Output
I expect 2 identical copies of the dataframe to be displayed.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 776fed3ab63d74ddef6e5af1a702b10c2a30bbb6
python: 3.7.0.final.0
python-bits: 64
OS: Darwin
OS-release: 17.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.0.dev0+418.g776fed3ab
pytest: 3.7.1
pip: 10.0.1
setuptools: 39.0.1
Cython: 0.28.5
numpy: 1.15.0
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: 1.7.6
patsy: None
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None
None

</details>

### PR DESCRIPTION
**Pull request by NasaGeek**
- [x] closes #22199
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by jbrockmendel — CHANGES_REQUESTED** [review]
Looks fine, needs a test.

**Comment by jreback** [comment]
can you merge master and update

**Comment by NasaGeek** [comment]
Rebased and expanded a test

File: pandas/tests/io/parser/test_read_fwf.py [hunk 1]
```diff
@@ -206,15 +206,17 @@ def test_fwf_compression(self):
...
-                tmp = compresser(path, mode='wb')
-                tmp.write(data)
-                tmp.close()
-                result = read_fwf(path, widths=widths, names=names,
-                                  compression=comp_name)
-                tm.assert_frame_equal(result, expected)
+        comps = [('gzip', 'gz', gzip.GzipFile), ('bz2', 'bz2', bz2.BZ2File)]
+        for comp_name, extension, compresser in comps:
+            for kwargs in [
+                    {'compression': comp_name}, {'compression': 'infer'}, {}]:
```
Comment (jreback) [inline pandas/tests/io/parser/test_read_fwf.py]: this needs to be parameterized.

@gfyoung don't we have functions to help compression testing along?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you merge master and update according to fixtures as pointed to by @gfyoung

**Comment by jreback** [comment]
@NasaGeek can you merge master and update

**Comment by jreback** [comment]
@NasaGeek can you merge master (or @gfyoung if you can)

**Comment by NasaGeek** [comment]
Happy to rebase, @jreback, but unfortunately I don't yet have an approach to your parametrization request. The available fixture doesn't quite give me enough to work with by itself. It looks like pandas.util.testing.decompress_file may provide a solution if it's augmented to allow return writable file objects. The issue there is that the open function in python2's zipfile module will not give back writable files. Would you be willing to miss coverage on zipfile in python2? If so, what's the best way to skip it?

**Comment by gfyoung** [comment]
@NasaGeek : You raise some good points regarding the fixtures.  I'm going to first do some refactoring on the `test_read_fwf.py` file itself first before revisiting this PR.

**Comment by NasaGeek** [comment]
Just wanted to confirm you didn't need anything else from me on this PR. @gfyoung's changes look fine to me.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/io/parser/test_read_fwf.py [hunk 2]
```diff
@@ -556,19 +556,24 @@ def test_default_delimiter():
...
 
 @pytest.mark.parametrize("compression", ["gzip", "bz2"])
-def test_fwf_compression(compression):
+@pytest.mark.parametrize("infer", [True, False, None])
+def test_fwf_compression(compression, infer):
     data = """1111111111
     2222222222
     3333333333""".strip()
 
+    extension = "gz" if compression == "gzip" else "bz2"
```
[the remark this answers is not in the document]
Comment (gfyoung) [reply pandas/tests/io/parser/test_read_fwf.py]: `compression_only` will pass in `gzip` for GZip files.  However, the file extension is `gz`, not `gzip`.  That's what we also follow in `_infer_compression` in `pandas/io/common`.

Thus, we have to do this.

[the remark this answers is not in the document]
Comment (gfyoung) [reply pandas/tests/io/parser/test_read_fwf.py]: Is `gzip` considered a valid file extension for GZip files?  `gz` is generally the most common, and is what we follow in `_infer_compression` in `pandas/io/common`.

File: pandas/tests/io/parser/test_read_fwf.py [hunk 3]
```diff
@@ -556,19 +556,24 @@ def test_default_delimiter():
...
     if compat.PY3:
         data = bytes(data, encoding="utf-8")
 
-    with tm.ensure_clean() as path:
+    with tm.ensure_clean(filename="tmp." + extension) as path:
         tm.write_to_compressed(compression, path, data)
 
-        result = read_fwf(path, compression=compression, **kwargs)
+        if infer is not None:
+            kwargs["compression"] = ("infer" if infer else compression)
```
[the remark this answers is not in the document]
Comment (gfyoung) [reply pandas/tests/io/parser/test_read_fwf.py]: Fixed.

File: pandas/tests/io/parser/test_read_fwf.py [hunk 4]
```diff
@@ -556,19 +556,24 @@ def test_default_delimiter():
 
 
 @pytest.mark.parametrize("compression", ["gzip", "bz2"])
```
[the remark this answers is not in the document]
Comment (gfyoung) [reply pandas/tests/io/parser/test_read_fwf.py]: Done.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping on green. @jbrockmendel if you have comments.

**Comment by gfyoung** [comment]
@jreback : Tests are all green.  PTAL.

@jbrockmendel : Any other thoughts?

**Review by mroeschke — APPROVED** [review]

**Pull request merged**